File: drv/general.c
int InitSaveData();
int InitSaveError();
int InitSaveE();
void Initialize();
int SaveData();
int CloseSaveData();
int CloseSaveError();
int CloseSaveE();
int GenerateMatlab(char *prefix);
void GetMass(KPP_REAL CL[], KPP_REAL Mass[]);
void INTEGRATE(double TIN, double TOUT);
double Stopwatch(double t);

int main()
{
  KPP_REAL dval[NSPEC];
  int i;
  double t;
  /* ---- TIME VARIABLES ------------------ */

  RTOLS = 1e-3;
  TSTART = 3600 * 12;
  TEND = TSTART + 3600 * 24 * 5;
  DT = 3600.;
  TEMP = 236.21;

  t = Stopwatch(0);
  // for (int j = 0; j < 10000; j++)
  // {
    Initialize();
    for (i = 0; i < NVAR; i++)
    {
      RTOL[i] = RTOLS;
      ATOL[i] = 1.0;
    }
    STEPMIN = 0.01;
    STEPMAX = 900;

    /* ********** TIME LOOP **************************** */

    InitSaveData();
    InitSaveError();
    InitSaveE();

    printf("\n%7s %7s   ", "done[%]", "Time[h]");
    for (i = 0; i < NMONITOR; i++)
      printf("%8s  ", SPC_NAMES[MONITOR[i]]);
    for (i = 0; i < NMASS; i++)
      printf("(%6s)  ", SMASS[i]);

    TIME = TSTART;
    FUNTIME = (double)TSTART;
    while (TIME <= TEND)
    {
      GetMass(C, dval);
      printf("\n%6.1f%% %7.2f   ", (TIME - TSTART) / (TEND - TSTART) * 100, TIME / 3600);
      for (i = 0; i < NMONITOR; i++)
        printf("%9.3e  ", C[MONITOR[i]] / CFACTOR);
      for (i = 0; i < NMASS; i++)
        printf("%9.3e  ", dval[i] / CFACTOR);

      SaveData();

      INTEGRATE(TIME, TIME + DT);
      TIME += DT;
      FUNTIME += (double)DT;
    }

    /* *********** END TIME LOOP *********************** */
  // }
  t = Stopwatch(t);
  printf("Simulation Time: %10.6f", t);
  printf("\n");
  CloseSaveData();
  CloseSaveError();
  CloseSaveE();

  return 0; /*didnt return anything initially */
}

File: drv/general.f90
PROGRAM KPP_ROOT_Driver

  USE KPP_ROOT_Model
  USE KPP_ROOT_Initialize, ONLY: Initialize

#ifdef __MIXEDPREC
      REAL(kind=dp) :: T
#else
      KPP_REAL :: T
#endif
      KPP_REAL :: DVAL(NSPEC)
      KPP_REAL :: RSTATE(20)
      INTEGER :: i
  
!~~~> Initialization 

      STEPMIN = 0.0d0
      STEPMAX = 0.0d0

      DO i=1,NVAR
        RTOL(i) = 1.0d-3
        ATOL(i) = 1.0
      END DO
     
      CALL Initialize()
      CALL InitSaveData()

!~~~> Time loop
      T = TSTART
#ifdef __MIXEDPREC
      FUNTIME = REAL(TSTART,kind=dp)
#endif      
kron: DO WHILE (T < TEND)

        TIME = T
        CALL GetMass( C, DVAL )
        WRITE(6,991) (T-TSTART)/(TEND-TSTART)*100, T,       &
                   ( TRIM(SPC_NAMES(MONITOR(i))),           &
                     C(MONITOR(i))/CFACTOR, i=1,NMONITOR )
        CALL SaveData()
        CALL Update_SUN() 
        CALL Update_RCONST()

        CALL INTEGRATE( TIN = T, TOUT = T+DT, RSTATUS_U = RSTATE, &
        ICNTRL_U = (/ 0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0 /) )
        T = RSTATE(1)
#ifdef __MIXEDPREC
      FUNTIME = FUNTIME + REAL(DT,kind=dp)
#endif

      END DO kron
!~~~> End Time loop

      CALL GetMass( C, DVAL )
      WRITE(6,991) (T-TSTART)/(TEND-TSTART)*100, T,     &
               ( TRIM(SPC_NAMES(MONITOR(i))),           &
                 C(MONITOR(i))/CFACTOR, i=1,NMONITOR ) 
      TIME = T
      CALL SaveData()
      CALL CloseSaveData()

991   FORMAT(F6.1,'%. T=',E9.3,2X,200(A,'=',E11.4,'; '))

END PROGRAM KPP_ROOT_Driver

